src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/DPBasedJoinPlanOptimizer.java

```diff
@@ -0,0 +1,164 @@
+                            candidatePlans.add( PhysicalPlanFactory.createPlanWithIndexNLJ( newRoot, plan_right ) );
+
+                            final FederationMember fm = ( (LogicalOpRequest<?, ?>)plan_left.getRootOperator() ).getFederationMember();
+                            if ( fm instanceof SPARQLEndpoint ){
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinFILTER( newRoot, plan_right ) );
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinUNION( newRoot, plan_right ) );
+                                candidatePlans.add( PhysicalPlanFactory.createPlanWithBindJoinVALUES( newRoot, plan_right ) );
+                            }
+                        }
+                    }
+
+                    // Prune: only the best candidate plan is retained in optPlan.
+                    // TODO: Move the cost annotation out of this for-loop. For all plans of the same size, invoke the cost function once.
+                    final List<PhysicalPlanWithCost> candidatesWithCost = PhysicalPlanWithCostUtils.annotatePlansWithCost(ctxt.getCostModel(), candidatePlans);
+                    final PhysicalPlan planWithLowestCost = PhysicalPlanWithCostUtils.findPlanWithLowestCost(candidatesWithCost).getPlan();
+                    optPlan.add( plans, planWithLowestCost );
+                }
+            }
+            return optPlan.get( subplans );
+        }
+
+    }
+
+    // This method returns all subsets (with the given size) of the given superset.
+    public static <T> List<List<T>> getSubSet( final List<T> superset, final int n ){
+        final List<List<T>> result = new ArrayList<>();
+        if( n==0 ){
+            result.add( new ArrayList<>() );
+            return result;
+        }
+
+        final List<List<T>> tempList = new ArrayList<>();
+        tempList.add( new ArrayList<>() );
+
+        for ( T element : superset ) {
+            final int size = tempList.size();
+            for ( int j = 0; j < size; j++ ){
+                // Stop adding more elements to the subset if its size is more than n.
+                if( tempList.get(j).size() >= n ){
+                    continue;
+                }
+
+                // Create a copy of the current subsets, then update the copy by adding an another element.
+                final List<T> clone = new ArrayList<>( tempList.get(j) );
+                clone.add( element );
+                tempList.add(clone);
+
+                // Add the subset to the result if its size is n.
+                if( clone.size() == n ){
+                    result.add(clone);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    // Split a superset into two subsets. This method returns all possible combinations of subsets.
+    public static <T> List<Pair<List<T>, List<T>>> splitIntoSubSets( final List<T> superset ){
```

Comment:
Just for the sake of safety, throw an `IllegalArgumentException` if `superset.size() < 2`.